### DISCUSSION TITLE
Why vmid is written twice in function tlb_vm_inv_all

### DISCUSSION THREAD

**Discussion by xioliu** [bao-project/bao-hypervisor#153] [General]
Hello,

May someone please explain why vmid is written twice here? Thanks.
static inline void tlb_vm_inv_all(asid_t vmid)
{
    uint64_t vttbr = 0;
    vttbr = sysreg_vttbr_el2_read();
    bool switch_vmid =
        bit64_extract(vttbr, VTTBR_VMID_OFF, VTTBR_VMID_LEN) != vmid;

    if (switch_vmid) {
        sysreg_vttbr_el2_write((((uint64_t)vmid << VTTBR_VMID_OFF) & VTTBR_VMID_MSK));
        DSB(ish);
        ISB();
    }

    arm_tlbi_vmalls12e1is();

    **if (switch_vmid) {
        DSB(ish);
        sysreg_vttbr_el2_write((((uint64_t)vmid << VTTBR_VMID_OFF) & VTTBR_VMID_MSK));
    }**
}

**Comment by josecm** [comment]
The reason for this is that the TLB invalidation instructions for the second stage assume the currently active VMID. If the current VMID is not the VMID we are trying to invalidate, we need to switch it before invalidating, and then switch back. Hope this helps!

**Reply by xioliu** [reply]
Here the vmid whose TLB to be invalidated is written twice.
As the old VMID be written for the 2nd time, code should be like this.

    arm_tlbi_vmalls12e1is(); 

    // Write back the original VMID if it was switched
    if (switch_vmid) {
        sysreg_vttbr_el2_write(vttbr);
    }

**Reply by DavidMCerdeira** [reply]
Nice catch!
If you can, please submit a PR with the fix.

**Reply by xioliu** [reply]
Will do.
